### PR TITLE
Initial KMS Service Operator

### PR DESCRIPTION
**Pull request by arajakul**
Issue #, if available:
#491 

Description of changes:
Initial ACK generation for KMS Service and smoke tests
* Includes initial tests for kms key creation
* Includes initial tests for kms grant

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by arajakul** [comment]
Please note above is a draft PR for early feedback. Do not merge 🙇

File: services/kms/pkg/resource/key/sdk.go [hunk 1]
```diff
@@ -0,0 +1,373 @@
...
+		if awsErr, ok := ackerr.AWSError(respErr); ok && awsErr.Code() == "UNKNOWN" {
+			return nil, ackerr.NotFound
+		}
+		return nil, respErr
+	}
+
+	// Merge in the information we read from the API call above to the copy of
+	// the original Kubernetes object we passed to the function
+	ko := r.ko.DeepCopy()
+
```
Comment (arajakul) [inline services/kms/pkg/resource/key/sdk.go:73]: I had to remove the following lines
```
-       if resp.KeyMetadata.CreationDate != nil {
-               ko.Status.CreationDate = &metav1.Time{*resp.KeyMetadata.CreationDate}
-       }
-       if resp.KeyMetadata.CustomerMasterKeySpec != nil {
-               ko.Spec.CustomerMasterKeySpec = resp.KeyMetadata.CustomerMasterKeySpec
-       }
-       if resp.KeyMetadata.Description != nil {
-               ko.Spec.Description = resp.KeyMetadata.Description
-       }
```

Reconcile was complaining about the diffs produced between desired state.
```
"diff": "\"Spec.CustomerMasterKeySpec\"\n\n\"Spec.KeyUsage\"\n\n\"Spec.Origin\"\n\n\"Status.CreationDate\"\n"
```

The heading of this file mentions that this file should not be edited, in which case where can I make some of these changes?

File: test/e2e/kms/smoke.sh [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+source "$SCRIPTS_DIR/lib/common.sh"
+source "$SCRIPTS_DIR/lib/k8s.sh"
+source "$SCRIPTS_DIR/lib/aws/$service_name.sh"
+source "$SCRIPTS_DIR/lib/testutil.sh"
+
+test_name="$( filenoext "${BASH_SOURCE[0]}" )"
+ack_ctrl_pod_id=$( controller_pod_id )
+debug_msg "executing test: $service_name/$test_name"
+key_name=ack-test-smoke-kms-413689294325
+#key_name="ack-test-smoke-$service_name-$AWS_ACCOUNT_ID"
```
Comment (arajakul) [inline test/e2e/kms/smoke.sh:17]:
```suggestion
key_name="ack-test-smoke-$service_name-$AWS_ACCOUNT_ID"
```

File: services/kms/pkg/resource/key/sdk.go (hunk 1 above)
Comment (jaypipes) [reply services/kms/pkg/resource/key/sdk.go:73]: @arajakul yeah, we're working on some changes to the code generator to handle certain field comparisons better. You've run into a great example of where the generated code falls down. You are correct that if you remove these lines of code, they will be re-inserted into `sdk.go` the next time you regenerate the controller. That said, there's nothing stopping you from making changes like this and committing the code; just be aware that if you regenerate, you'll overwrite those changes :)

Comment (arajakul) [reply services/kms/pkg/resource/key/sdk.go:73]: If you could point me to the code where those comparisons are taking place, I might be able to submit a fix.

**Comment by arajakul** [comment]
Ran some basic tests and looks like we are able to create kms key and associate a key with a kms grant without issues.

Let me know if there are other tests or changes you'd like to see

File: services/kms/pkg/resource/key/sdk.go (hunk 1 above)
Comment (jaypipes) [reply services/kms/pkg/resource/key/sdk.go:73]: Hi @arajakul! I'm working on the compare stuff in this branch: https://github.com/jaypipes/aws-controllers-k8s/tree/delta

It is dependent on the refactoring currently going on in the code generator here:

https://github.com/aws/aws-controllers-k8s/pull/620

I expect to have the refactoring merged today and complete the compare resource changes early next week.

File: test/e2e/kms/smoke.sh [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+debug_msg "checking grant creation for user $user in $service_name for kms key $key_id"
+aws_return=$(daws kms list-grants --key-id $key_id | jq -r '.Grants [] .GranteePrincipal')
+assert_equal "$user" "$aws_return" "Expected $user to have access to kms grant" || exit 1
+
+sleep 1
+
+kubectl delete "key/$resource_name" 2>/dev/null
+assert_equal "0" "$?" "Expected success from kubectl delete but got $?" || exit 1
+
+assert_pod_not_restarted $ack_ctrl_pod_id
```
Comment (jaypipes) [inline test/e2e/kms/smoke.sh:70]: :+1: nice test.

**Review by jaypipes — APPROVED** [review]
welcome to the ACK contributor club, @arajakul :)

**Pull request merged**